### PR TITLE
Add cider-load-all-files and clojure-files-in-directory functions

### PR DESCRIPTION
**Pull request by fromheten**
Useful when the nrepl host is on another machine, and you wish to "force" the
state of your local file system onto that remote nrepl all at once. 

For this use case `cider-load-all-project-ns` will not do as it runs on the nrepl server machine (unless I've misunderstood). 

-----------------

Before submitting the PR make sure the following things have been done (and denote this
by checking the relevant checkboxes):

- [x] The commits are consistent with our [contribution guidelines][1]
- [N/A] You've added tests (if possible) to cover your change(s)
- [x] All tests are passing (`make test`)
- [x] The new code is not generating bytecode or `M-x checkdoc` warnings
- [x] You've updated the changelog (if adding/changing user-visible functionality)
- [N/A] You've updated the readme (if adding/changing user-visible functionality)
- [N/A] You've updated the refcard (if you made changes to the commands listed there)

### CODE REVIEW

File: cider-interaction.el [hunk 1]
```diff
@@ -1611,9 +1611,56 @@ The heavy lifting is done by `cider-load-buffer'."
     (find-file filename)
     (cider-load-buffer (current-buffer))))
 
+(defun clojure-files-in-directory (directory)
```
Comment (bbatsov) [inline cider-interaction.el]: I'd opt for some much simpler implementation of this in terms of `directory-files` - e.g. https://github.com/rejeep/f.el/blob/master/f.el#L495

File: cider-interaction.el [hunk 2]
```diff
@@ -1611,9 +1611,56 @@ The heavy lifting is done by `cider-load-buffer'."
...
+  "List of all .clj files found recursively in DIRECTORY.
+
+Fork of function at: https://www.gnu.org/software/emacs/manual/html_node/eintr/Files-List.html"
+  (interactive "DDirectory name: ")
+  (cider-ensure-connected)
```
Comment (bbatsov) [inline cider-interaction.el]: You don't really need cider to be connected for this to work.

File: cider-interaction.el [hunk 3]
```diff
@@ -1611,9 +1611,56 @@ The heavy lifting is done by `cider-load-buffer'."
...
+  (let (clj-files-list
+        (current-directory-list
+         (directory-files-and-attributes directory t)))
+    ;; while we are in the current directory
+    (while current-directory-list
+      (cond
+       ;; check to see whether filename ends in '.clj'
+       ;; and if so, add its name to a list.
+       ((equal ".clj" (substring (car (car current-directory-list)) -4))
```
Comment (bbatsov) [inline cider-interaction.el]: Btw, I'd just get a listing of all the files and filter the results for clojure files.

File: cider-interaction.el [hunk 4]
```diff
@@ -1611,9 +1611,56 @@ The heavy lifting is done by `cider-load-buffer'."
     (find-file filename)
     (cider-load-buffer (current-buffer))))
 
+(defun clojure-files-in-directory (directory)
+  "List of all .clj files found recursively in DIRECTORY.
+
+Fork of function at: https://www.gnu.org/software/emacs/manual/html_node/eintr/Files-List.html"
+  (interactive "DDirectory name: ")
```
Comment (bbatsov) [inline cider-interaction.el]: I don't think this needs to be interactive at all.

File: cider-interaction.el [hunk 5]
```diff
@@ -1611,9 +1611,56 @@ The heavy lifting is done by `cider-load-buffer'."
...
+                 (clojure-files-in-directory
+                  (car (car current-directory-list)))
+                 clj-files-list)))))
+      ;; move to the next filename in the list; this also
+      ;; shortens the list so the while loop eventually comes to an end
+      (setq current-directory-list (cdr current-directory-list)))
+    ;; return the filenames
+    clj-files-list))
+
+(defun cider-load-all-files (directory)
```
Comment (bbatsov) [inline cider-interaction.el]: This should probably have an autoload cookie.

File: cider-interaction.el [hunk 6]
```diff
@@ -1611,9 +1611,56 @@ The heavy lifting is done by `cider-load-buffer'."
...
+  "Load all files in DIRECTORY (recursively)."
```
Comment (bbatsov) [inline cider-interaction.el]: I'd expand here a bit on the usefulness of this command.

File: cider-interaction.el [hunk 7]
```diff
@@ -1611,9 +1611,56 @@ The heavy lifting is done by `cider-load-buffer'."
...
+  (interactive "DDirectory name: ")
+  (mapcar #'cider-load-file
+          (clojure-files-in-directory directory)))
+
 (defalias 'cider-eval-file 'cider-load-file
   "A convenience alias as some people are confused by the load-* names.")
 
+(defalias 'cider-eval-all-files 'cider-load-file
```
Comment (bbatsov) [inline cider-interaction.el]: You made a copy/paste mistake here.

File: cider-interaction.el [hunk 8]
```diff
@@ -1611,9 +1611,56 @@ The heavy lifting is done by `cider-load-buffer'."
...
+            (equal "."
+                   (substring (car (car current-directory-list)) -1))
+            ;; then do nothing since filename is that of
+            ;;   current directory or parent, "." or ".."
+            ()
+          ;; else descend into the directory and repeat the process
+          (setq clj-files-list
+                (append
+                 (clojure-files-in-directory
+                  (car (car current-directory-list)))
```
Comment (xiongtx) [inline cider-interaction.el]: `caar`

**Comment by bbatsov** [comment]
I can see the usefulness of such a command. A few notes:

* this has to be added to the manual
* this has to be added to the menus

File: cider-interaction.el (hunk 1 above)
Comment (fromheten) [inline cider-interaction.el]: Absolutely right - I was unaware of `directory-files-recursively` so ended up with this shit there. Thanks for pointing it out, submitting a new commit in a moment with this procedure removed.

**Comment by fromheten** [comment]
@bbatsov I've submitted commits implementing your and @xiongtx s ideas, all very good. Now my procedure is just a 2-liner, which is way better. 

The keybinding has been added to menus and to the documentation. The only documentation I haven't compiled is the tex refcard as I don't have a LaTeX compiler set up. Let me know if you want me to verify that before being satisfied. 

If the code looks OK I'll rebase the commits into one so that they can be merged into `master`. 

Thank you for CIDER!

**Pull request merged**